File: app/src/main/java/com/weiqi/sgf/Sgf.kt
package com.weiqi.sgf

import com.weiqi.engine.GameState
import com.weiqi.engine.MoveType
import com.weiqi.engine.Point
import com.weiqi.engine.ScoreResult
import com.weiqi.engine.StoneColor
import java.time.LocalDate

object Sgf {

    /** Encode a board column index (0-based) to SGF letter (a, b, ...). SGF supports up to 25. */
    private fun coord(p: Point): String {
        val col = ('a' + p.col)
        val row = ('a' + p.row)
        return "$col$row"
    }

    fun export(
        state: GameState,
        score: ScoreResult? = null,
        blackName: String = "Black",
        whiteName: String = "White",
        date: LocalDate = LocalDate.now()
    ): String {
        val sb = StringBuilder()
        sb.append("(;GM[1]FF[4]CA[UTF-8]AP[Weiqi]")
        sb.append("SZ[").append(state.config.boardSize).append(']')
        sb.append("KM[").append(state.config.komi).append(']')
        sb.append("HA[").append(state.config.handicap).append(']')
        sb.append("RU[Chinese]")
        sb.append("PB[").append(blackName).append(']')
        sb.append("PW[").append(whiteName).append(']')
        sb.append("DT[").append(date).append(']')
        if (score != null) sb.append("RE[").append(score.resultString).append(']')

        for (m in state.history) {
            val tag = if (m.player == StoneColor.BLACK) "B" else "W"
            when (m.type) {
                MoveType.PLACE_STONE -> sb.append(';').append(tag).append('[').append(coord(m.point!!)).append(']')
                MoveType.PASS -> sb.append(';').append(tag).append("[]")
                MoveType.RESIGN -> { /* recorded via RE */ }
            }
        }
        sb.append(')')
        return sb.toString()
    }
}

File: lib/src/sgf/sgf.dart
import '../domain/game_state.dart';
import '../domain/models.dart';
import '../domain/scoring.dart';

class Sgf {
  /// SGF RU[] property tags. See https://www.red-bean.com/sgf/properties.html#RU.
  static String _sgfRuleset(Ruleset r) {
    switch (r) {
      case Ruleset.chinese:
        return 'Chinese';
      case Ruleset.japanese:
        return 'Japanese';
      case Ruleset.korean:
        return 'Korean';
      case Ruleset.aga:
        return 'AGA';
      case Ruleset.ing:
        return 'Ing';
      case Ruleset.newZealand:
        return 'NZ';
      case Ruleset.trompTaylor:
        return 'TrompTaylor';
    }
  }

  static String _sgfComment(GameConfig cfg) {
    final defaults = RulesetDefaults.of(cfg.ruleset);
    final scoring = defaults.scoringMethod == ScoringMethod.area
        ? 'area scoring'
        : 'territory + prisoners scoring';
    final ko = switch (defaults.superkoMode) {
      SuperkoMode.none => 'basic ko',
      SuperkoMode.positional => 'positional superko',
      SuperkoMode.situational => 'situational superko',
    };
    final suicide = defaults.allowSuicide ? ', suicide legal' : '';
    // SGF text values must escape ']' and '\\'. The strings below are static
    // and never include those characters, so no escaping is required.
    return '${cfg.ruleset.label} rules — $scoring, komi ${cfg.komi}, $ko$suicide.';
  }

  static Ruleset rulesetFromSgf(String s) {
    switch (s.trim().toLowerCase()) {
      case 'japanese':
        return Ruleset.japanese;
      case 'korean':
        return Ruleset.korean;
      case 'aga':
        return Ruleset.aga;
      case 'ing':
      case 'goe':
        return Ruleset.ing;
      case 'nz':
      case 'new zealand':
        return Ruleset.newZealand;
      case 'tromptaylor':
      case 'tromp-taylor':
      case 'tromp taylor':
        return Ruleset.trompTaylor;
      case 'chinese':
      default:
        return Ruleset.chinese;
    }
  }

  /// Encode a Point to SGF coordinate letters (col, row).
  static String _coord(Point p) {
    final col = String.fromCharCode('a'.codeUnitAt(0) + p.col);
    final row = String.fromCharCode('a'.codeUnitAt(0) + p.row);
    return '$col$row';
  }

  static String _escape(String value) =>
      value.replaceAll('\\', '\\\\').replaceAll(']', '\\]');

  static String export(
    GameState state, {
    ScoreResult? score,
    String? result,
    String blackName = 'Black',
    String whiteName = 'White',
    DateTime? date,
  }) {
    final d = date ?? DateTime.now();
    final dStr =
        '${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';
    final sb = StringBuffer();
    sb.write('(;GM[1]FF[4]CA[UTF-8]AP[Go]');
    sb
      ..write('SZ[')
      ..write(state.config.boardSize)
      ..write(']');
    sb
      ..write('KM[')
      ..write(state.config.komi)
      ..write(']');
    sb
      ..write('HA[')
      ..write(state.config.handicap)
      ..write(']');
    sb
      ..write('RU[')
      ..write(_sgfRuleset(state.config.ruleset))
      ..write(']');
    sb
      ..write('PB[')
      ..write(_escape(blackName))
      ..write(']');
    sb
      ..write('PW[')
      ..write(_escape(whiteName))
      ..write(']');
    sb
      ..write('DT[')
      ..write(dStr)
      ..write(']');
    result = result?.isNotEmpty == true ? result : score?.resultString;
    if (result == null &&
        state.status == GameStatus.resigned &&
        state.history.isNotEmpty) {
      result = '${state.history.last.player.other.short}+R';
    }
    if (result != null) {
      sb
        ..write('RE[')
        ..write(_escape(result))
        ..write(']');
    }
    // Human-readable game comment summarizing the ruleset. Most SGF viewers
    // surface GC[] alongside the player names, so the ruleset is visible
    // at-a-glance without inspecting the raw properties.
    sb
      ..write('GC[')
      ..write(_escape(_sgfComment(state.config)))
      ..write(']');

    if (state.config.handicap > 0) {
      final initial = GameState.newGame(state.config).board;
      sb.write('AB');
      for (var row = 0; row < initial.size; row++) {
        for (var col = 0; col < initial.size; col++) {
          final point = Point(row, col);
          if (initial.cellAt(point) == CellState.black) {
            sb.write('[${_coord(point)}]');
          }
        }
      }
      sb.write('PL[W]');
    }
    for (final m in state.history) {
      final tag = m.player == StoneColor.black ? 'B' : 'W';
      switch (m.type) {
        case MoveType.placeStone:
          sb
            ..write(';')
            ..write(tag)
            ..write('[')
            ..write(_coord(m.point!))
            ..write(']');
          break;
        case MoveType.pass:
          sb
            ..write(';')
            ..write(tag)
            ..write('[]');
          break;
        case MoveType.resign:
          // Recorded via RE[]; no move token.
          break;
      }
    }
    sb.write(')');
    return sb.toString();
  }
}
